docs/design/hooks.md

```diff
@@ -0,0 +1,322 @@
+
+	// ActionRemoveOnCreateVolumeEvent represent remove action on volume create Event
+	ActionRemoveOnCreateVolumeEvent ActionType = "removeEntriesOnCreateVolumeEvent"
+
+	// ActionAddOnDeleteVolumeEvent represent add action on volume delete Event
+	ActionAddOnDeleteVolumeEvent ActionType = "addOrUpdateEntriesOnDeleteVolumeEvent"
+
+	// ActionRemoveOnDeleteVolumeEvent represent remove action on volume delete Event
+	ActionRemoveOnDeleteVolumeEvent ActionType = "removeEntriesOnDeleteVolumeEvent"
+)
+
+// PVHook defines the field which will be updated for PV Hook Action
+type PVHook struct {
+	// Annotations needs to be added/removed on/from the PV
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the PV
+	Finalizers []string `json:"finalizers,omitempty"`
+}
+
+// PVCHook defines the field which will be updated for PVC Hook Action
+type PVCHook struct {
+	// Annotations needs to be added/removed on/from the PVC
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the PVC
+	Finalizers []string `json:"finalizers,omitempty"`
+}
+
+// ServiceHook defines the field which will be updated for Service Hook Action
+type ServiceHook struct {
+	// Annotations needs to be added/removed on/from the Service
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the Service
+	Finalizers []string `json:"finalizers,omitempty"`
+}
+
+// DeploymentHook defines the field which will be updated for Deployment Hook Action
+type DeploymentHook struct {
+	// Annotations needs to be added/removed on/from the Deployment
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Finalizers needs to be added/removed on/from the Deployment
+	Finalizers []string `json:"finalizers,omitempty"`
+}
+
+// HookConfig represent the to be executed by nfs-provisioner
+type HookConfig struct {
+	// Name represent hook name
+	Name string `json:"name"`
+
+	// NFSPVConfig represent config for NFSPV resource
+	NFSPVConfig *PVHook `json:"nfsPV,omitempty"`
+
+	// BackendPVConfig represent config for BackendPV resource
+	BackendPVConfig *PVHook `json:"backendPV,omitempty"`
+
+	// BackendPVCConfig represent config for BackendPVC resource
+	BackendPVCConfig *PVCHook `json:"backendPVC,omitempty"`
+
+	// NFSServiceConfig represent config for NFS Service resource
+	NFSServiceConfig *ServiceHook `json:"nfsService,omitempty"`
+
+	// NFSDeploymentConfig represent config for NFS Deployment resource
+	NFSDeploymentConfig *DeploymentHook `json:"nfsDeployment,omitempty"`
+
+	// Event defines provisioning event type on which
+	// given hook action needs to be executed
+	Event EventType `json:"eventType"`
+
+	// Action represent the type of hook action, i.e HookActionAdd or HookActionRemove
+	Action HookActionType `json:"actionType"`
```

Comment:
Do we need this? Since in `Hook` struct it is map of actionOnEventType --> HookConfig IMO it is not required to add again in HookConfig